### PR TITLE
fix the device type for with_comms decorator

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125798

found by @yifuwang, it looks like we are wrongly using
self.device_type="cuda" for gloo backend, which are triggering some
flakiness. i.e. https://github.com/pytorch/pytorch/issues/125366

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 1]
```diff
@@ -359,11 +355,11 @@ def with_comms(func: TestFunc) -> TestFunc:
     def wrapper(
         self, *args: Tuple[object], **kwargs: Dict[str, Any]  # type: ignore[misc]
     ) -> None:
-        # if backend not specified, and cuda available, then use nccl, else gloo
-        if torch.cuda.is_available() and torch.cuda.device_count() >= self.world_size:
-            self.device_type = "cuda"
-        else:
+        # if enough GPU we can use GPU, otherwise we fallback to CPU
+        if torch.cuda.is_available() and torch.cuda.device_count() < self.world_size:
             self.device_type = "cpu"
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:360]: I understand your intention, but the "and" here seems a bit weird. 
What about:
```
if not torch.cuda.is_available() or torch.cuda.device_count() < self.world_size:
```
Or maybe I missed your other intention.

Comment (wanchaol) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py:360]: Sure that works too, let me change to that

Comment (wanchaol) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py:360]: this might also fix the current CI issues, let's see

**Review by yifuwang — APPROVED** [review]
LG if CI passes!

**Pull request closed without merging**